### ISSUE TITLE
Test, improve and write documentation of VirtualBox user image

### ISSUE THREAD

**Issue by sapetnioc** [brainvisa/casa-distro#141]
There is a first [VirutalBox user image for the master branch of brainvisa distro](http://brainvisa.info/casa-distro/releases/brainvisa-5.0-test.ova). This image can be imported in VirtualBox in a few clicks. During this procedure, a few image parameters can be choosen (e.g. memory size, number of CPUs). Then it can be started and it will be the starting point for many users of BrainVISA.

I propose this issue to discuss about all bugs, improvements and documentation regarding this VM. This could replace the system specific issues (#84, #85 and #86) with regards of VirtualBox.

Here is a list (that we can edit) of what should work on Linux, Windows and Mac to close this issue: 

- [ ] Find installation doc
- [ ] Install VM in VirtualBox
- [ ] Start VM
- [ ] Find how to start in the desktop environment
- [ ] Use Anatomist
- [ ] Use BrainVISA
- [ ] Use Aims commands
- [ ] Find doc to share a directory
- [ ] Share a directory with test data
- [ ] Perform a Morphologist analysis with BrainVISA and/or MorphologistUI

**Comment by sapetnioc** [comment]
I had no problem to import the VM and start BrainVISA and Anatomist on my Linux laptop (VirtualBox 6.1.12 r139181).

**Comment by denisri** [comment]
What's the state of 3D and GPU with virtualbox ?

**Comment by sapetnioc** [comment]
Since guest additions are installed, user should be able to enable 3D
acceleration in VM properties (a reboot is required). This have to be
tested and documented.

Le lun. 21 sept. 2020 à 16:29, Denis Rivière <notifications@github.com> a
écrit :

**Comment by sapetnioc** [comment]
Bad news. My first try on windows : `brainvisa` is starting with a lot of errors in the console and `anatomist` segfault` :

```
brainvisa@brainvisa-VirtualBox:~$ anatomist
QApplication: invalid style override passed, ignoring it.
Starting Anatomist.....
config file : /home/brainvisa/.anatomist/config/settings.cfg
PyAnatomist Module present
Traceback (most recent call last):
  File "/usr/lib/python2.7/site.py", line 68, in <module>
    import os
  File "/usr/lib/python2.7/os.py", line 400, in <module>
    import UserDict
  File "/usr/lib/python2.7/UserDict.py", line 116, in <module>
    import _abcoll
  File "/usr/lib/python2.7/_abcoll.py", line 70, in <module>
    Iterable.register(str)
  File "/usr/lib/python2.7/abc.py", line 109, in register
    if issubclass(subclass, cls):
  File "/usr/lib/python2.7/abc.py", line 174, in __subclasscheck__
    for rcls in cls._abc_registry:
  File "/usr/lib/python2.7/_weakrefset.py", line 60, in __iter__
    for itemref in self.data:
TypeError: iter() returned non-iterator of type 'setiterator'
Segmentation fault (core dumped)
```

**Comment by sapetnioc** [comment]
By default, there is only 8Mb of video memory allocated to VM. Raising it to 50 Mb prevents the crash of Anatomist.

**Comment by sapetnioc** [comment]
On Windows I have problem with 3D acceleration. If I enable it (using  VMSVGA driver as recommended by VirtualBox), the graphical interface cannot start with 50 Mb video RAM (infinite loop when X session is staring). With 128Mb, it starts but it is impossible to open a terminal (seems to be stuck). But if 3D acceleration is selected, VirtualBox force the VMSVGA driver. Using VMSVGA without 3D acceleration seems to work.

**Comment by sapetnioc** [comment]
I tried to re-install guest addition drivers on Windows. It removed the 6.1.12 release to replace it by 6.1.4 but it changed nothing regarding terminal failure with 3D acceleration.

**Comment by denisri** [comment]
The error in iter() looks weird: how can python work differently in a VM ?

**Comment by sapetnioc** [comment]
There's no error when there is enough video memory. My Linux laptop works
with 8 Mb but have a lower resolution and smaller screen than my Windows
one.

**Comment by sapetnioc** [comment]
This is worse than I expected. I do not know if we have a problem or if I have a problem with VirtualBox on Windows but I was not able to install Ubuntu nor Lubuntu this morning. The procedure always finished by the machine being stuck or crashed.

**Comment by denisri** [comment]
I have tried the virtualbox image on Mac. I ended up with something more or less working, but it was kind of a pain:

* The VM consistently crashed when booting linux, no matter the memory, CPU, video settings. It took me some time to have the idea to look at the logs and realize that it always crashed after logging a message about audio detection. Setting audio to "fake audio device" finally fixed the problem. How could I have known ? ;)
* graphics display only worked for a small screen size, until I largely increased the video memory (to 64 Mo), otherwise enlarging the VM window lead to a black screen.
* Fonts and icons are microscopic, it's almost impossible to read anything. I had to scale the display in virtualbox to 200% (but then I guess we only have half the real resolution - unless it's a fake resolution as Apple sometimes does ?)
* Graphics display is slow, really slow, even in 2D, even when 3D acceleration is enabled.
* 3D acceleration is a joke, it's very slow whatever the display settings in virtualbox. Enabling 3D acceleration only makes volume rendering in anatomist render a black cube.
* Mac keyboad is not correctly recognized and layed out, some keys do not produce the expected character. I couldn't find a keybload layout setting in lubuntu...
* Mac mouse doesn't have a middle button and doesn't seem to have an "emulation mode", so I could not use rotations in anatomist until I plugged a "real" mouse (which hopefully works). I think I remember I did some special modifs in anatomist code to use the mouse differently on Mac, but here we are in linux mode...

**Comment by denisri** [comment]
Frankly the singularity option on Mac seems simpler and works far better for me, once we solve the env variables issue. No memory / cpu / video / audio hassle to configure "blindly" before it works, keyboard settings are the host one thus work as the user expects, and even display (either 2D or 3D with software mesa) looks far smoother / faster and with correct icons/fonts sizes. The mouse issue is the same thus, I think we should setup some settings or env variable to tell anatomist to get into 2-button mouse mode, it's worth the effort.

**Comment by denisri** [comment]
Plus, the singularity images are way smaller.

**Comment by sapetnioc** [comment]
The bad news : I failed with VirtualBox and Ubuntu 20.04 on another Windows machine
The good news : I did volume rendering with Anatomist on Windows using a free version of Xming (its old since the project had become comercial).

So I propose to completely drop VirtualBox support for now and put all our effort on Singularity on all Systems. However, we are lacking a solution to do accelerated display on Mac.
